thefuck/rules/nix_shell.py

```diff
@@ -0,0 +1,40 @@
+from thefuck.specific.nix import nix_available
+import subprocess
+
+enabled_by_default = nix_available
+
+# Set the priority just ahead of `fix_file` rule, which can generate low quality matches due
+# to the sheer amount of paths in the nix store.
+priority = 999
+
+
+def get_nixpkgs_name(bin):
+    """
+    Returns the name of the Nix package that provides the given binary. It uses the
+    `command-not-found` binary to do so, which is how nix-shell generates it's own suggestions.
+    """
+
+    result = subprocess.run(
+        ["command-not-found", bin], stderr=subprocess.PIPE, universal_newlines=True
+    )
+
+    # return early if package is not available through nix
+    if "nix-shell" not in result.stderr:
+        return ""
+
+    nixpkgs_name = result.stderr.split()[-1] if result.stderr.split() else ""
+    return nixpkgs_name
+
+
+def match(command):
+    bin = command.script_parts[0]
+    return (
+        "nix-shell" not in command.script
+        and "command not found" in command.output
```

Comment:
In your example:

```
$ ponysay moo
The program 'ponysay' is not in your PATH. You can make it available in an
ephemeral shell by typing:
  nix-shell -p ponysay

$ fuck
nix-shell -p ponysay --run "ponysay moo" [enter/↑/↓/ctrl+c]
```

`command not found` is not part of the output. Didn't you mean:
```suggestion
        and "command not found" not in command.output
```

Reply:
> In your example, command not found is not part of the output

For commands that could be made available through nix:

1. Indeed, the visible output in the terminal does *not* contain the text "command not found".

```bash
$ spt
The program 'spt' is not in your PATH. You can make it available in an
ephemeral shell by typing:
  nix-shell -p spotify-tui
```

1. But when we log the value of `command` to file, we do see "command not found":

```
Command(script=spt, output=/nix/store/p6dlr3skfhxpyphipg2bqnj52999banh-bash-5.2-p15/bin/sh: line 1: spt: command not found)
```

I'm not sure exactly why they're different, but it works 🤷🏼‍♂️.

Admittedly, it could be simpler if we could do:

```diff
- and "command not found" in command.output
+ and command.exitcode is 127
```

But I couldn't find a way to access the _numeric_ exit code (127) from within the `match` function. Only the human [error message](https://tldp.org/LDP/abs/html/exitcodes.html) (command not found) is available it seems.

Finally, I've annotated the match conditions to make it clear what my intentions are for each.

Reply:
Ah, yes. That's because the command is executed through the shell — `/bin/sh`. Would you please run thefuck in debug mode and look for the line that says `DEBUG: Received output:`? e.g.:

```shell
THEFUCK_DEBUG=true thefuck spt
```

Among the debug messages there will be one with the output generated by the shell. Could you then please update the output values in `mocked_nixpkgs`?

Also, how about writing [tests for `match`](https://github.com/nvbn/thefuck/blob/62e0767c5069aeee176b0fe3459068b7703aaa26/tests/rules/test_brew_update_formula.py#L10-L19)? 😊

Reply:
> tests for match

done

I've cleaned up the PR a fair bit. As for the mock output, I've indicated which command I've used to get them so their purpose should be more clear now. If you notice that I've missed anything let me know. Thank you.